Add support for OpenTelemetry in sqlcommenter-sequelize

adamegyed opened the pull request.

odeke-em commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/index.js:
```diff
@@ -12,29 +12,44 @@
     'traceparent': false,
 };
 
+const providerMap = {
```
We shouldn't have to set this up in the package itself. Let's move this into provider.js

odeke-em commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/provider/opencensusprovider.js:
```diff
@@ -0,0 +1,28 @@
+const {tracer} = require('@opencensus/nodejs');
+const {toW3CTraceContext} = require('../util');
+
+exports.OpenCensusProvider = class OpenCensusProvider {
```
Quite verbose for a name, I think OpenCensus will suffice given that it is in a package already called provider, thus provider.OpenCensus is obvious.

odeke-em commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/provider/opentelemtryprovider.js:
```diff
@@ -0,0 +1,40 @@
+// Copyright 2019 Google LLC
```
Please use 2020 for the copyright year since that's when this code is being added.
s/2019/2020/g

Also the file name spells "opentelemtry" instead of "opentelemetry", please update that, and we can remove the provider.js suffix since it is verbose.

odeke-em commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/provider/opentelemtryprovider.js:
```diff
@@ -0,0 +1,40 @@
+const {trace, TraceFlags} = require('@opentelemetry/api');
+const {DEFAULT_INSTRUMENTATION_PLUGINS} = require('@opentelemetry/node/build/src/config');
+
+exports.OpenTelemetryProvider = class OpenTelemetryProvider {
```
exporters.OpenTelemetry

odeke-em commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/test/comment.test.js:
```diff
@@ -59,7 +63,7 @@ describe("Comments for Sequelize", () => {
 
     after(() => {
-        tracing.stop();
+        opencensus_tracing.stop();
```
We don't need to manually test opencensus_tracing any more given that we'll be abstracting it into a package that can be multiplexed on by name.

adamegyed replied:
This was to rename "tracing" to make it more obvious which library it was interacting with. The previous tracing test is still fully functional and passing, so I wasn't sure about removing them.

odeke-em commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/provider/index.js:
```diff
@@ -0,0 +1,19 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const {OpenCensusProvider} = require('./opencensusprovider');
+const {OpenTelemetryProvider} = require('./opentelemtryprovider');
+
+exports.OpenCensusProvider = OpenCensusProvider;
+exports.OpenTelemetryProvider = OpenTelemetryProvider;
```
I don't think we need to export OpenCensusProvider nor OpenTelemetryProvider, we can just create the logic in here
to multiplex on the various providers thus
```javascript
function attachComments(providerName, comments) {
     // Lookup the provider by name, or use the default.
     provider = providers[providerName] || OpenCensus;
}
```

odeke-em commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/traceprovider.js:
```diff
@@ -0,0 +1,23 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * TraceProvider describes the expected fields of a framework-specific trace provider.
+ * Currently, this is implemented by OpenCensusProvider and OpenTelemetryProvicder 
+ */
+exports.TraceProvider = class TraceProvider {
+    getW3CTraceContext() {
+        return {};
+    }
+};
```
Is this code needed?

odeke-em commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/index.js:
```diff
@@ -12,29 +12,44 @@
  * @return {void}
  */
-exports.wrapSequelize = (sequelize, include={}) => {
+exports.wrapSequelize = (sequelize, include={}, options={TraceProvider: 'OpenCensus'}) => {
```
Why 'OpenCensus' and why not 'opencensus' and 'opentelemetry'? CamelCase for a string option might be too tough on the eyes for users who can type 'Opencensus' or 'Opentelemetry' accidentally.

adamegyed replied:
I think I'll make the config accept any capitalization of the name to simplify things.

odeke-em commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/index.js:
```diff
@@ -58,9 +73,9 @@ exports.wrapSequelize = (sequelize, include={}) => {
-        }
+        // Add trace context to comments, depending on the current provider.
+        const traceContext = traceProvider.getW3CTraceContext();
+        Object.assign(comments, traceContext);
```
I think we should abstract this setup into the code in provider.js or even util.js itself thus, in  the endhere, all that we'll need to do inside this file is perhaps
```javascript
provider.attachComments(options.TraceProvider, comments);
```

kintel commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/package.json:
```diff
@@ -7,13 +7,16 @@
     "@opencensus/propagation-tracecontext": "^0.0.15",
+    "@opentelemetry/core": "^0.10.1",
+    "@opentelemetry/api": "^0.10.1",
+    "@opentelemetry/node": "^0.10.2",
```
This is a question to node.js experts: Why are these devDependencies? Should these rather be peer dependencies? How do we avoid importing both opencensus and opentelemetry if we're using only one?

kintel commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/provider/opencensusprovider.js:
```diff
@@ -0,0 +1,28 @@
+            const carrier = {};
+            toW3CTraceContext(tracer.currentRootSpan, carrier);
+            return carrier;
+        } else {
```
To make this slightly more readable, could you pull both the carried declaration and return value out of the if-else, and eliminate the else?

kintel commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/provider/opentelemtryprovider.js:
```diff
@@ -0,0 +1,40 @@
+            }
+        });
+
+        const traceParent = `00-${spanContext.traceId}-${spanContext.spanId}-0${
```
This would need to be kept in sync with the upstream OpenTelemetry repo, and is a maintenance risk. Could we use OpenTelemetry's own propagator to inject this header into an object?

kintel commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/provider/opentelemtryprovider.js:
```diff
@@ -0,0 +1,40 @@
+        let spanContext = null;
+        // Collect a list of the default loaded nodejs tracers.
+        // The tracer names are required to retrieve the latest span.
+        let validPluginNames = Object.keys(DEFAULT_INSTRUMENTATION_PLUGINS)
```
This feels like a rather cumbersome way of extracting the current trace context. As virtually every integration like this would need similar functionality, have you discussed this with the OpenTelemetry team?

kintel commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/test/comment.test.js:
```diff
@@ -160,19 +164,19 @@ describe("Excluding all variables", () => {
 });
 
-describe("With tracing", () => {
+describe("With OpenCensus tracing", () => {
```
As mentioned above, we should also test for the default case. ..and if keeping the two levels of default handling, test against an unknown provider string as well.

odeke-em replied:
Ack, please take a look at @kintel's suggestion here.

adamegyed replied:
Added

adamegyed commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/index.js:
```diff
@@ -12,29 +12,44 @@
+    }
+    let traceProvider = providerMap[options.TraceProvider];
+    if (!traceProvider) {
+        traceProvider = OpenCensusProvider;
```
I fixed this and reworked it in my update to be clearer. And as for disabling tracing, I think the expected behavior of old clients is to keep appending trace data through OpenCensus, so that option should be non-default. Any thoughts @odeke-em ?

adamegyed commented:
I've made some updates to the PR, addressing almost all of the comments. Here's a full list of what I changed:

 - Removed the TraceProvider "interface"
 - Renamed providers to just the library names, referenced from provider's index.js
 - Updated copyrights to 2020
 - Fixed naming issue with OpenTelemetryProvider (now losing the provider)
 - Provider index now does the multiplexing, as well as holding the string->provider mapping
 - Updated the string->provider mapping to disregard capitalization
 - Updated the logic of how provider works, and adjust the multi-line interaction in the index.js file to this:
`provider.attachComments(options.TraceProvider, comments);`
 - Fixed how the default Provider is set to OpenCensus.
 - Update OpenTelemetryProvider to use their injection logic rather than reimplementing


The last issue was super tricky, and after many different attempts, it turns out this simple solution was the correct way to do it all along. The only issue is that it depends on a yet-to-be-published change in @opentelemetry/context-base, addressing the issue described here: https://github.com/open-telemetry/opentelemetry-js/issues/1379. The issue is actually a little bit deeper, as the previous version's use of Symbol meant that all invocations of `getActiveSpanContext(...)` and `getParentSpanContext(...)` would fail when run from a different directory than where @opentelemetry/core was first `require(...)` 'ed. 

What I've committed works with this change in my local testing, but we just have to wait until the next version update for `@opentelemetry/context-base` (which should be this week) before merging.

I still have outstanding questions regarding whether we should support a no-trace-data-commenting option, and for whether we should move the tracing dependencies into peer dependencies or regular dependencies.

kintel commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/provider/opentelemetry.js:
```diff
@@ -0,0 +1,23 @@
+
+exports.OpenTelemetry = class OpenTelemetry {
+    static addW3CTraceContext(comments) {
+        let propagator = new HttpTraceContext();
```
FYI: OpenTelemetry offers a global propagator, which is configurable by customers: https://github.com/open-telemetry/opentelemetry-js/blob/master/packages/opentelemetry-api/src/api/propagation.ts

I'm not sure what the correct approach is here. Perhaps the CloudSQL team can comment on what propagation they expect or support on the back-end.

adamegyed replied:
I also tried that approach before, but I avoided it due to the side effect of having to adjust the global default propagator. This is what it would look like:
```javascript
propagation.setGlobalPropagator(new HttpTraceContext());
propagation.inject(context.active(), comments, defaultSetter);
```
Without modifying the global propagator, we would be unsure of what method of propagation would be used, which could be something other than the Cloud SQL's expected W3C Trace Context. If the user hasn't set one before, then calling inject will default to a Noop propagator.

odeke-em commented on nodejs/sqlcommenter-nodejs/packages/sqlcommenter-sequelize/provider/index.js:
```diff
@@ -0,0 +1,30 @@
+
+exports.attachComments = function attachComments(providerName, comments) {
+    // Verify we have a comments object to modify
+    if (!comments || !(typeof(comments) === 'object')) return;
```
`typeof comments !== 'object'`

adamegyed replied:
got it

adamegyed commented:
Thanks @odeke-em ! There's just one more thing I want to try out, and it's regarding @kintel 's suggestion of not requiring both imports in order to use sqlcommenter. I'm going to try moving the tracing libraries to peer dependencies and dynamically loading the libraries during the wrapSequelize step, which should be alright given that it only runs once per service start and at initial load.

odeke-em commented:
Personally I’d advise that as a separate step, please file an issue then
work on that after the core part is implemented, otherwise you’ll be
blocked until that is all fixed which could take forever. We still have
knex.js to also implement which moves the needle forward and the other work
will be a vestige :)

On Wed, Aug 12, 2020 at 3:10 PM Adam Egyed <notifications@github.com> wrote:

> Thanks @odeke-em <https://github.com/odeke-em> ! There's just one more
> thing I want to try out, and it's regarding @kintel
> <https://github.com/kintel> 's suggestion of not requiring both imports
> in order to use sqlcommenter. I'm going to try moving the tracing libraries
> to peer dependencies and dynamically loading the libraries during the
> wrapSequelize step, which should be alright given that it only runs once
> per service start and at initial load.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/google/sqlcommenter/pull/14#issuecomment-673135311>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABFL3V6O5TQOPYGU5AKTLETSAMHMHANCNFSM4PL3HJ4Q>
> .
>

adamegyed commented:
Good point. In that case it should be good to go!

odeke-em commented:
Done, thank you @kintel for the review and congrats @adamegyed for the change!
